src/ba.c

```diff
@@ -1,11 +1,41 @@
 #include "gsa.h"
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
 
 void border_array(size_t n, const char x[n], int ba[n])
 {
     // FIXME
+    int c = 0, i = 1;
+    int len = strlen(x);
+    ba[0] = 0;
+    while (i < len) {
+        if (x[c] == x[i]) {
+            c++;
+            ba[i] = c;
+            i++;
```

Comment:
If we want to show off our crazy C skilzz then this would be `ba[i++] = ++c;`. With `++c` we increment first and return the new value, and with `i++` we get the current value first and then increment.